codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/endpoints/EndpointRulesEngineGenerator.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen.endpoints;
+
+import java.util.Optional;
+import software.amazon.smithy.go.codegen.SmithyGoDependency;
+import software.amazon.smithy.go.codegen.SymbolUtils;
+import software.amazon.smithy.go.codegen.integration.ProtocolGenerator;
+import software.amazon.smithy.rulesengine.language.EndpointRuleSet;
+import software.amazon.smithy.rulesengine.traits.EndpointRuleSetTrait;
+
+
+
+/**
+ * Generates all components required for Smithy Ruleset Endpoint Resolution.
+ * These components include a Provider, Parameters, and Tests.
+ */
+public class EndpointRulesEngineGenerator {
+
+    public static final String FEATURE_NAME = "RulesEngine";
+    public static final String PARAMETERS_TYPE_NAME = "EndpointParameters";
+    public static final String RESOLVER_INTERFACE_NAME = "Endpoint" + FEATURE_NAME + "Resolver";
```

Comment:
yah so the rationale behind this is the following:
-  `EndpointResolver` is already taken (it was the initial public type that exists currently)
- `EndpointProvider` i didnt use the term "provider" anywhere rather i used "resolver". This is because ultimately the existing Go interface we want to comply with has its function named `ResolveEndpoint` and my understanding is that "resolving an endpoint" and "providing an endpoint" is the same thing (correct me if im wrong). so it felt confusing to see both of these terms as unintentionally different ways of describing the same thing.
- i didnt see "rules engine" as a description of an implementation but rather the public facing name of "endpoints 2.0", this is because this name is used in all base smithy packages for endpoints 2.0 stuff.

Reply:
discussed offline, going to go with `V2` instead of `RulesEngine` because a customer may implement this themselves without rules and V2 signifies that this is an iteration and should be used over the previous.